docs/ADR-016-Backwards-and-Forwards-Compatibility.md

```diff
@@ -0,0 +1,136 @@
+
+
+Replace era-variant GADTs (like Certificate era) with a uniform wrapper (Certificate :: L.TxCert era -> Certificate era) and using open type families to resolve era-specific types (e.g., Delegatee era). Era constraints (IsEra, IsShelleyBasedEra) gate access to functionality.
+
+I will make this clearer with an example below.
+
+**NB**: We will only implement functionality constrained by `IsShelleyBasedEra era` and `IsEra era` that either is currently exposed by the "old" api or specifically required by QA. If we do not need backwards compatibility then we will use`Era era` or `IsEra era` without implementing an additional type synonym family.
+
+
+Let's use `makeStakeAddressDelegationCertificate` as an example.
+
+```haskell
+data StakeDelegationRequirements era where
+  StakeDelegationRequirementsConwayOnwards
+    :: ConwayEraOnwards era
+    -> StakeCredential
+    -> Ledger.Delegatee
+    -> StakeDelegationRequirements era
+  StakeDelegationRequirementsPreConway
+    :: ShelleyToBabbageEra era
+    -> StakeCredential
+    -> PoolId
+    -> StakeDelegationRequirements era
+
+makeStakeAddressDelegationCertificate :: StakeDelegationRequirements era -> Certificate era
+makeStakeAddressDelegationCertificate = \case
+  StakeDelegationRequirementsConwayOnwards cOnwards scred delegatee ->
+    conwayEraOnwardsConstraints cOnwards $
+      ConwayCertificate cOnwards $
+        Ledger.mkDelegTxCert (toShelleyStakeCredential scred) delegatee
+  StakeDelegationRequirementsPreConway atMostBabbage scred pid ->
+    shelleyToBabbageEraConstraints atMostBabbage $
+      ShelleyRelatedCertificate atMostBabbage $
+        Ledger.mkDelegStakeTxCert (toShelleyStakeCredential scred) (unStakePoolKeyHash pid)
+
+-- The `StakeDelegationRequirements era` GADT will necessitate more data constructors if there are any changes to the requirements of stake delegation in future eras. 
+-- This is already seen in the transition from Babbage (can only delegate to stake pools) -> Conway (can delegate to stake pools and dreps). 
+-- This obviously does not scale in the face of possible incoming changes and requires the propagation of an additional data constructor.
+
+-- I propose the following:
+
+-- Already exists in Cardano.Api.Experimental.Tx.Internal.Certificate
+data Certificate era where
+  Certificate :: L.EraTxCert era => L.TxCert era -> Certificate era
+
+-- Type family definition capturing the change of possible delegatees. We can now delegate to stake pools or dreps or delegate
+-- to dreps and stake pools simultaneously. This is captured in the `Ledger.Delegatee` type. 
+type family Delegatee era where
+  Delegatee ConwayEra = Ledger.Delegatee
+  Delegatee BabbageEra = Api.Hash Api.StakePoolKey
+  Delegatee AlonzoEra = Api.Hash Api.StakePoolKey 
+  ... 
+  ...
+  Delegate ShelleyEra = Api.Hash Api.StakePoolKey
```

Comment:
```suggestion
  Delegatee L.ConwayEra = Ledger.Delegatee
  Delegatee L.BabbageEra = Api.Hash Api.StakePoolKey
  Delegatee L.AlonzoEra = Api.Hash Api.StakePoolKey 
  ... 
  ...
  Delegate L.ShelleyEra = Api.Hash Api.StakePoolKey
```

`Delegatee` is a part of the experimental (new) API so I think we should be using ledger eras.